README.md
![logo](./media/logo.png)

> Bots navigating urls and doing tasks.

`damon` is a tool that runs on [CasperJS](http://casperjs.org/) which runs on [PhantomJS](http://phantomjs.org/).

It feeds on JSON files that describe what tasks he needs to achieve on specified starting URL.

# Table Of Contents
<details>

<!-- toc -->

- [Damon Projects](#damon-projects)
- [Installation](#installation)
- [Usage](#usage)
  * [Locally](#locally)
  * [CLI](#cli)
- [Task File](#task-file)
  * [config](#config)
  * [tasks](#tasks)
    + [navigate](#navigate)
    + [status](#status)
    + [redirection](#redirection)
    + [capture](#capture)
    + [download](#download)
    + [wait](#wait)
      - [url](#url)
      - [selector](#selector)
      - [visible](#visible)
      - [hidden](#hidden)
      - [time](#time)
      - [resource](#resource)
    + [dom](#dom)
      - [click](#click)
      - [fill](#fill)
    + [get](#get)
      - [_store_](#_store_)
        * [attribute](#attribute)
        * [variable](#variable)
        * [resource](#resource-1)
        * [number of elements](#number-of-elements)
      - [_access_](#_access_)
    + [request](#request)
    + [assert](#assert)
      - [attribute](#attribute-1)
      - [variable](#variable-1)
      - [key](#key)
- [Roadmap](#roadmap)
- [Contribute](#contribute)
      - [Individual Contribution](#individual-contribution)
      - [Corporate Contribution](#corporate-contribution)
- [License](#license)

<!-- tocstop -->

</details>

## Damon Projects

- [💻 CLI](https://github.com/damonjs/damon-cli).
- [💬 Reporter](https://github.com/damonjs/damon-reporter).

## Installation

via NPM :

```bash
npm install --save damon
```

## Usage

### Locally

```node
var damon = require('damon');
// Attach the default reporter.
damon.attachReporter();
// You can attach your own reporter as well
// damon.attachReporter('./path/to/my/reporter.js');
// Start your suite(s), it accepts globs.
damon.start('./tasks.json');
```

### CLI

You can use `damon` via a CLI, available at [damonjs/damon-cli](https://github.com/damonjs/damon-cli)

## Task File

This is the json file you'll pass to `damon`.

It's composed of two attributes, a `config` hash and a `tasks` array.

```javascript
{
    "config": {},
    "tasks": []
}
```

### config

Your task file must have a `config` entry with a `size` and a `url`.

```javascript
"config": {
    "size": {
        "width": 1024,
        "height": 768
    },
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) AppleWebKit/537.75.14 (KHTML, like Gecko) Version/7.0.3 Safari/7046A194A",
    "timeout": 1000,
    "logLevel": "fatal",
    "describe": "This is a job description"
}
```

- `size` is for the viewport's size.
- `userAgent` is a custom userAgent for `damon` to use. Default one is Chrome 44 on Windows 7.
- `timeout` overwrite the general timeout used accross the test suite.
- `logLevel` control at which level `damon` will log. Can be `none`, `fatal`, `error`, `warn`, `info`, `debug` or `trace`
- `describe` is used to give a description of the job. It is printed next to the filename in the default reporter.

### tasks

Then you describe your tasks in a `tasks` entry that is an array of all the tasks to achieve sequentially :

```javascript
"tasks": [
    {},
    {}
]
```

Each task will have three components:

```javascript
{
    "type": "taskType",
    "it": "should run this task",
    "params": {}
}
```

- `type` give the type of task for `damon` to run.
- `params` provide arguments to be passed to the task.
- `it` provide a description of the task, to be printed on the default reporter (optional).

__It exists several kinds of tasks that `damon` can achieve :__

#### navigate

`damon` can navigate to other urls at the start or during its worflow.

```javascript
{
    "type": "navigate",
    "params": {
        "url": "https://www.google.com",
        "method": "GET",
        "data": {
            "key": "value"
        },
        "headers": {
            "name": "value"
        },
        "encoding": "utf8"
    }
}
```

Only `params.url` is required.

#### status

Verify that the page answers with a specific status.

```javascript
{
    "type": "status",
    "params": {
        "url": "url",
        "method": "GET",
        "data": {
            "key": "value"
        },
        "headers": {
            "name": "value"
        },
        "encoding": "utf8",
        "status": [301, 302]
    }
}
```

`damon` will navigate to `params.url` and will wait until it gets the awaited status.

The request gets cancelled as soon as the status is returned.

`params.status` can be a string of a single status, or an array of all status required.

In the case of the array, the first encounter will validate the task.

Only `params.url`and `params.status` are required.

#### redirection

Verify that the page redirects to another specified one.

```javascript
{
    "type": "redirection",
    "params": {
        "from": "url",
        "to": "url",
        "method": "GET",
        "data": {
            "key": "value"
        },
        "headers": {
            "name": "value"
        },
        "encoding": "utf8"
    }
}
```

`damon` will navigate to `params.from` and verify that we have a redirection to `params.to`.

The `params.method`, `params.data`, `params.heades` and `params.encoding` are for `params.from` only.

Only `params.from`and `params.to` are required.

#### capture

A simple screen capture :

```javascript
{
    "type": "capture",
    "params": {
        "name": "start.png"
    }
}
```
#### download

Download the target url

```javascript
{
    "type": "download",
    "params": {
        "url": "http://www.google.com",
        "name": "google.html",
        "method": "GET",
        "data": ""
    }
}
```

An HTTP method can be set with `method`, and pass request arguments through `data`.

#### wait

`damon` can wait for several different things.
For each one, except `time`, you can overwrite the `timeout`.

##### url

```javascript
{
    "type": "wait",
    "params": {
        "url": "http://www.yahoo.ca",
        "regexp": false,
        "timeout": 1000
    }
}
```

`damon` will wait at this step until matching url is reached.

`url` will be interpreted as a `regexp` if set to `true`. Default value of `regexp` is `false`.

##### selector

```javascript
{
    "type": "wait",
    "params": {
        "selector": "#content",
        "timeout": 1000,
        "xpath": false
    }
}
```

`damon` will wait at this step until the `selector` is available on the page.

`xpath` can be used to select an element by setting it to true. Default value is false.

##### visible
##### hidden

Both are the same as `selector` but will wait for these specific states of the element.

##### time

```javascript
{
    "type": "wait",
    "params": {
        "time": "1000"
    }
}
```

`damon` will wait for the specified amount of milliseconds.

##### resource

```javascript
{
    "type": "wait",
    "params": {
        "resource": "resourceName",
        "regexp": false,
        "timeout": 1000,
        "method": "DELETE"
    }
}
```

`damon` will wait at this step until something matching the resource is received.

`resource` will be interpreted as a `regexp` if set to `true`. Default value of `regexp` is `false`.

A `method` can be specified to filter the resource. If nothing is specified, any `method` will be accepted.

#### dom

`damon` can perform two different actions on a dom element :

##### click

```javascript
{
    "type": "dom",
    "params": {
        "selector": "button#btnSubmit",
        "xpath": false,
        "do": "click",
    }
}
```

`damon` will click on the specified selector.

##### fill

```javascript
{
    "type": "dom",
    "params": {
        "selector": "input#userName",
        "xpath": false,
        "do": "fill",
        "text": "yoann.dev"
    }
}
```

`damon` will enter text in the specified field.

`xpath` cannot be used when filling a file field due to [PhantomJS limitiations](http://docs.casperjs.org/en/latest/modules/casper.html#fill).

#### get

##### _store_

`damon` can perform different `get` to retrieve a value and store it for subsequent tasks :

###### attribute

```javascript
{
    "type": "get",
    "params": {
        "selector": "div#Info",
        "xpath": false,
        "attribute": "title",
        "key": "infoTitle",
        "modifier": "[a-z]+"
    }
}
```

`damon` will get the value of the `attribute`, apply the `modifier` RegExp and store it as `infoTitle`.

`@text` can also be used as an `attribute` to get the text content of the `selector`

###### variable

```javascript
{
    "type": "get",
    "params": {
        "variable": "var.attr1['attr2']",
        "key": "varAttr2"
    }
}
```

`damon` will access to the specified variable with `window` as the root object and store its value as `varAttr2`

###### resource

```javascript
{
    "type": "get",
    "params": {
        "resource": "resourceLink",
        "regexp": false,
        "variable": "payload.title",
        "key": "title",
        "method": "POST"
    }
}
```

`damon` will access to the specified variable of the matching `resource` and store it.

A `method` can be specified to filter the resource. If nothing is specified, any `method` will be accepted.

To access to a variable in the payload of a resource, write `payload.variableName` for `variable` field. Resource also contains the `headers`, `method`, `time` and `url`.

###### number of elements

```javascript
{
    "type": "get",
    "params": {
        "selector": "ul#list li",
        "xpath": false,
        "key": "liNumber"
    }
}
```

`damon` will store the number of elements that satisfy the `selector`

##### _access_

The value can then be accessed in any following tasks via its `key` value

```javascript
{
    "type": "wait",
    "params": {
        "url": "http://www.yahoo.ca/{{key}}"
    }
}
```

To access the stored value, call the `key` in between double brackets `{{key}}`

#### request

`damon` can perform any HTTP call.

```javascript
{
    "type": "request",
    "params": {
        "url": "https://www.google.com",
        "method": "GET",
        "payload": {
            "q": "funny cats"
        },
        "headers": {
            "header": "value"
        },
        "store": {
            "key": "key",
            "variable": "variable.attr1"
        }
    }
}
```

You can also `store` the response for later use with `{{key}}`.

If you don't pass a `variable` it will store the complete response.

Otherwise, it will try to parse the response as JSON and look for your variable.

#### assert

`damon` can perform different `assert` actions to test a value with an expected value:

##### attribute

```javascript
{
    "type": "assert",
    "params": {
        "selector": "div#Info",
        "xpath": false,
        "attribute": "title",
        "modifier": "[a-z]+",
        "expected": "expectedValue or {{key}}"
    }
}
```

`damon` will `get` the value of the `attribute` and test it against the `expected` value or the value associated with `{{key}}`

##### variable

```javascript
{
    "type": "assert",
    "params": {
        "variable": "var.attr1['attr2']",
        "expected": "expectedValue or {{key}}"
    }
}
```

`damon` will `get` the value of the `variable` and test it against the `expected` value or the value associated with `{{key}}`

##### key

```javascript
{
    "type": "assert",
    "params": {
        "key": "title",
        "expected": "Expected Title"
    }
}
```

`damon` will `get` the value of the `key` and test it against the `expected` value.

## Roadmap

- [x] Extract the CLI component.
- [x] Extract the default reporter component.
- [ ] Extract default actions. (place them in the config file to be imported)
- [ ] Extract default plugins. (place them in the config file to be imported)
- [ ] Create Sitemap crawling plugin (almost there).
- [ ] Create Retry mecanic plugin.
- [ ] Create Github Page to use as main website.
- [ ] Create wiki to document how to create reporters/actions/plugins.
- [ ] Extract runner component.
- [ ] Extract logger component.
- [ ] Add examples.
- [ ] Test everything individually.

## Contribute

We welcome Your interest in Autodesk’s Open Source Damon (the “Project”).

Any Contributor to the Project must accept and sign an Agreement indicating agreement
to the license terms below.

##### [Individual Contribution](http://goo.gl/forms/ctQNFrveEF)
##### [Corporate Contribution](http://goo.gl/forms/4DTn9ho2JT)

## License

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.

You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
<a name="0.6.2"></a>
## [0.6.2](https://github.com/damonjs/damon/compare/v0.6.0...v0.6.2) (2016-08-16)


### Features

* **docs:** add a ToC ([38e4859](https://github.com/damonjs/damon/commit/38e4859))



<a name="0.6.0"></a>
# [0.6.0](https://github.com/damonjs/damon/compare/v0.5.15...v0.6.0) (2016-08-15)


### Bug Fixes

* **deps:** update dependencies ([81a3dcd](https://github.com/damonjs/damon/commit/81a3dcd))
* **runner:** force parse log after each suite, not only end ([13ccf7b](https://github.com/damonjs/damon/commit/13ccf7b))


### Features

* **actions:** use http instead of the navigate standalone ([e728cc9](https://github.com/damonjs/damon/commit/e728cc9))
* **assert:** rename assert.url to assert.status for better semantics ([8fed25e](https://github.com/damonjs/damon/commit/8fed25e))
* **http:** add an action set for http stuff ([507c792](https://github.com/damonjs/damon/commit/507c792))
* **reporter:** upgrade to damon-reporter 0.1.0 ([0e32886](https://github.com/damonjs/damon/commit/0e32886))
* **runner:** make the log parser synchronous ([45bca38](https://github.com/damonjs/damon/commit/45bca38))
* **worker:** remove the prepended task of navigate ([ff9afbe](https://github.com/damonjs/damon/commit/ff9afbe))



<a name="0.5.15"></a>
## [0.5.15](https://github.com/damonjs/damon/compare/v0.5.14...v0.5.15) (2016-07-07)


### Features

* **assert:** add a regexp flag for assert.status ([e59d534](https://github.com/damonjs/damon/commit/e59d534))



<a name="0.5.14"></a>
## [0.5.14](https://github.com/damonjs/damon/compare/v0.5.13...v0.5.14) (2016-07-06)



<a name="0.5.13"></a>
## [0.5.13](https://github.com/damonjs/damon/compare/v0.5.12...v0.5.13) (2016-07-06)


### Features

* **reporter:** put the default reporter in its own repo ([9b0d895](https://github.com/damonjs/damon/commit/9b0d895))



<a name="0.5.12"></a>
## [0.5.12](https://github.com/damonjs/damon/compare/v0.5.11...v0.5.12) (2016-07-05)


### Features

* **runner:** make a file watch compatible with network file system (eg. docker) ([5273d83](https://github.com/damonjs/damon/commit/5273d83))



<a name="0.5.11"></a>
## [0.5.11](https://github.com/damonjs/damon/compare/v0.5.10...v0.5.11) (2016-07-05)



<a name="0.5.10"></a>
## [0.5.10](https://github.com/damonjs/damon/compare/v0.5.9...v0.5.10) (2016-07-05)



<a name="0.5.9"></a>
## [0.5.9](https://github.com/damonjs/damon/compare/v0.5.8...v0.5.9) (2016-07-05)


### Bug Fixes

* **runner:** better reporting ([653d793](https://github.com/damonjs/damon/commit/653d793))



<a name="0.5.8"></a>
## [0.5.8](https://github.com/damonjs/damon/compare/v0.5.7...v0.5.8) (2016-07-05)



<a name="0.5.7"></a>
## [0.5.7](https://github.com/damonjs/damon/compare/v0.5.6...v0.5.7) (2016-07-05)



<a name="0.5.6"></a>
## [0.5.6](https://github.com/damonjs/damon/compare/v0.5.5...v0.5.6) (2016-07-05)



<a name="0.5.5"></a>
## [0.5.5](https://github.com/damonjs/damon/compare/v0.5.4...v0.5.5) (2016-07-05)



<a name="0.5.4"></a>
## [0.5.4](https://github.com/damonjs/damon/compare/v0.5.3...v0.5.4) (2016-07-05)


### Bug Fixes

* **reporter:** check for object before logging it ([fe33076](https://github.com/damonjs/damon/commit/fe33076))



<a name="0.5.3"></a>
## [0.5.3](https://github.com/damonjs/damon/compare/v0.5.2...v0.5.3) (2016-07-04)


### Bug Fixes

* **navigate:** clean listeners to avoid leaks and handle load fails ([5fc73dd](https://github.com/damonjs/damon/commit/5fc73dd))
* **navigate:** remove undefined requests usage ([965f541](https://github.com/damonjs/damon/commit/965f541))
* **runner:** escape double-quotes one more time ([26eafa7](https://github.com/damonjs/damon/commit/26eafa7))


### Features

* **worker:** add the resourceTimeout setting ([a0e2c04](https://github.com/damonjs/damon/commit/a0e2c04))



<a name="0.5.2"></a>
## [0.5.2](https://github.com/damonjs/damon/compare/v0.5.1...v0.5.2) (2016-07-04)


### Bug Fixes

* **start, navigate:** add a timeout to handle long requests ([f9a7dfa](https://github.com/damonjs/damon/commit/f9a7dfa))
* better error reporting on navigation ([de84fab](https://github.com/damonjs/damon/commit/de84fab))


### Features

* **casper:** add a flag to ignore errors ([01995fd](https://github.com/damonjs/damon/commit/01995fd))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/damonjs/damon/compare/v0.5.0...v0.5.1) (2016-06-29)


### Bug Fixes

* **reporter:** update the display of the report ([69f73a7](https://github.com/damonjs/damon/commit/69f73a7))
* **worker:** add missing error type ([ca35e58](https://github.com/damonjs/damon/commit/ca35e58))


### Features

* **runner:** attach the error object to the error report byError ([e4207f2](https://github.com/damonjs/damon/commit/e4207f2))
* **runner:** ignore 'Operation canceled' errors ([816b56b](https://github.com/damonjs/damon/commit/816b56b))
* **runner:** pass the type into the error event and task object ([33ec590](https://github.com/damonjs/damon/commit/33ec590))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/damonjs/damon/compare/v0.4.4...v0.5.0) (2016-06-29)


### Bug Fixes

* **worker:** consider the first navigate as a regular task ([4a74307](https://github.com/damonjs/damon/commit/4a74307))


### Features

* **reporter:** consider errors in the CLI ([55aaa4b](https://github.com/damonjs/damon/commit/55aaa4b))
* **reporter:** display a report at the end ([2ed8c9e](https://github.com/damonjs/damon/commit/2ed8c9e))
* **reporter:** display times in seconds ([37d76a2](https://github.com/damonjs/damon/commit/37d76a2))
* **runner:** create a report at the end of the run ([efa0796](https://github.com/damonjs/damon/commit/efa0796))
* **runner:** emit an 'error' event ([78d67ea](https://github.com/damonjs/damon/commit/78d67ea))
* **worker:** add a type to errors ([815aa47](https://github.com/damonjs/damon/commit/815aa47))
* **worker:** report both resource and javascript errors ([95793d1](https://github.com/damonjs/damon/commit/95793d1))



<a name="0.4.4"></a>

src/bots/worker.js
var currentFile = require('system').args[3];
var fs = require('fs');
var dirname = fs.absolute(currentFile).split('/');
// Needs to be loaded to be available across the lifecycle.
var utils = require('utils');
var system = require('system');
var pid = system.pid;

// Pass the PhantomJS pid to the runner
console.log('PhantomJS PID: ' + pid);

var opts, currentTask;

if (dirname.length > 1) {
    dirname.pop();
    dirname = dirname.join('/');
    fs.changeWorkingDirectory(dirname);
}

var casper = require('casper').create({
    clientScripts: ['./includes/start.js'],
    pageSettings: {
        loadImages: true,
        loadPlugins: true,
        localToRemoteUrlAccessEnabled: true
    },
    requests: {},
    verbose: false,
    exitOnError: false,
    // Overwrite logs to avoid having them in the console.
    onDie: function (casp, message, status) {
        log('casper died', message, status, 'ERROR');
    },
    onError: function (err, stack) {
        log('casper error', err, stack, 'ERROR');
    },
    onAlert: function (message) {
        log('casper alert', message, 'WARNING');
    },
    onStepTimeout: function _onStepTimeout(timeout, stepNum) {
        log(
            'Maximum step execution timeout exceeded for step ' + stepNum,
            'ERROR'
        );
    },
    onTimeout: function _onTimeout(timeout) {
        log('Script timeout of ' + timeout + ' reached, exiting.', 'ERROR');
    },
    onWaitTimeout: function _onWaitTimeout(timeout) {
        log('Wait timeout of ' + timeout + ' expired, exiting.', 'ERROR');
    },
    // Store all requests sent by casper into the casper object
    onResourceRequested: function (casp, requestData, networkRequest) {
        casper.options.requests[requestData.id] = requestData;
    }
});

// removing default options passed by the Python executable
casper.cli.drop('cli');
casper.cli.drop('casper-path');

if (!casper.cli.has('tasks')) {
    casper
        .echo('worker n°' + pid + ' needs tasks... will shut down.')
        .exit();
}

try {
    opts = require(casper.cli.get('tasks'));
} catch (err) {
    console.log('No tasks found', err);
    casper.exit();
}

var tasks = opts.tasks;
var cwd = casper.cli.has('cwd') ? casper.cli.get('cwd') : dirname;
var config = opts.config;
var logLevel = config.logLevel !== undefined ?
    config.logLevel : 'none';
var log = require('./log').config.call(casper, pid, logLevel);
var logger = require('./logger')(cwd);
var userAgent = config.userAgent || 'Mozilla/5.0 (Windows NT 6.1; WOW64) ' +
    'AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.157 Safari/537.36';

casper.options.waitTimeout = config.timeout !== undefined ?
    config.timeout : 10000;
casper.options.logLevel = logLevel;
casper.options.pageSettings.resourceTimeout = casper.options.waitTimeout;
casper.options.viewportSize = config.size ?
    config.size : {
        width: 1024,
        height: 720
    };
casper.userAgent(userAgent);

var actions = require('./actions').config.call(casper, cwd);

function startTask (task) {
    casper.options._ignoreErrors = false;
    task.start = +new Date();
    task.logId = logger.write(task, 'TASK.START');
    return task;
}

function configEndTask (task) {
    task.end = +new Date();
    task.duration = task.end - task.start;
    return task;
}

function errorTask (task, message, details, type) {
    logger.write({
        error: message,
        details: details,
        logId: task.logId,
        type: type
    }, 'TASK.ERROR');
}

function failTask (task) {
    logger.write(configEndTask(task), 'TASK.FAIL');
}

function endTask (task) {
    logger.write(configEndTask(task), 'TASK.END');
}

// GO.
casper.start();
// The Loop.
casper.then(function () {
    tasks.forEach(function (task) {
        casper.then(function () {
            currentTask = task;
            startTask(currentTask);
            try {
                return actions.execute(task);
            } catch (e) {
                log('Catched', e.message, e, 'ERROR');
                errorTask(currentTask, 'thrown error : ' + e.message,
                    e, 'task');
                failTask(currentTask);
            }
        }).then(function () {
            // Close the current task.
            endTask(currentTask);
        });
    });
});

casper.on('error', function(msg, backtrace) {
    log('error', arguments, 'ERROR');
    if (currentTask) {
        errorTask(currentTask, msg, backtrace, 'default');
    }
    failTask(currentTask);
});
casper.on('step.error', function(err) {
    log('step.error', arguments, 'ERROR');
    errorTask(currentTask, 'step error : ' + err.message, err, 'step');
    failTask(currentTask);
});
casper.on('step.timeout', function(step, timeout) {
    log('step.timeout', arguments, 'ERROR');
    errorTask(currentTask, 'timeout step', {
        step: step,
        timeout: timeout
    }, 'timeout');
    failTask(currentTask);
});
casper.on('resource.error', function (msg) {
    log('Resource Error', msg, 'ERROR');
    if (!casper.options._ignoreErrors) {
        errorTask(currentTask, msg.errorString, msg, 'resource');
    }
});
casper.on('page.error', function (msg, trace) {
    log('Js Error', msg, 'ERROR');
    if (!casper.options._ignoreErrors) {
        errorTask(currentTask, msg, trace, 'js');
    }
});
casper.on('timeout', function() {
    log('timeout', arguments, 'ERROR');
    errorTask(currentTask, 'timeout', null, 'timeout');
    failTask(currentTask);
});
casper.on('waitFor.timeout', function(timeout, params) {
    log('waitFor.timeout', arguments, 'ERROR');
    errorTask(currentTask, 'timeout waitFor', {
        timeout: timeout,
        params: params
    }, 'timeout');
    failTask(currentTask);
});

casper.run();
